File: frontend/App/Models/Post.swift
//
// Copyright © 2019 Sunho Kim. All rights reserved.
//

import Foundation

struct Post: Codable{
    var id: Int = 0
    var userId: Int = 0
    var createdAt: Date = Date()
    var updatedAt: Date = Date()
    var bookId: Int = 0
    var topContent: String = ""
    var sentence: String = ""
    var bottomContent: String = ""
    var solved: Bool = false
    var solvingContent: String?
    var solvingComment: Int?
    var rate: Int?
    var commentCount: Int?
    
    enum CodingKeys: String, CodingKey
    {
        case id
        case userId = "user_id"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
        case bookId = "book_id"
        case topContent = "top_content"
        case sentence
        case bottomContent = "bottom_content"
        case solved
        case solvingContent = "solving_content"
        case solvingComment = "solving_comment"
        case rate
        case commentCount = "comment_count"
    }
}

struct Comment: Codable {
    var id: Int = 0
    var userId: Int = 0
    var postId: Int = 0
    var createdAt: Date = Date()
    var updatedAt: Date = Date()
    var content: String = ""
    var rate: Int? = nil
    
    enum CodingKeys: String, CodingKey
    {
        case id
        case userId = "user_id"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
        case postId = "post_id"
        case content
        case rate
    }
}

File: backend/apiserver/models/dbmodels/post.go
//
// Copyright © 2019 Sunho Kim. All rights reserved.
//

package dbmodels

import (
	"time"

	"github.com/gobuffalo/nulls"
	"github.com/gofrs/uuid"
)

type Post struct {
	ID        int       `db:"id" json:"id"`
	UserID    int       `db:"user_id" json:"user_id"`
	CreatedAt time.Time `db:"created_at" json:"created_at"`
	UpdatedAt time.Time `db:"updated_at" json:"updated_at"`
}

type PostComment struct {
	ID        int       `db:"id" json:"id"`
	UserID    int       `db:"user_id" json:"user_id"`
	PostID    int       `db:"post_id" json:"post_id"`
	CreatedAt time.Time `db:"created_at" json:"created_at"`
	UpdatedAt time.Time `db:"updated_at" json:"updated_at"`
	Content   string    `db:"content" json:"content"`
}

type SentencePost struct {
	ID             int          `db:"id" json:"id"`
	BookID         int          `db:"book_id" json:"book_id"`
	TopContent     string       `db:"top_content" json:"top_content"`
	Sentence       string       `db:"sentence" json:"sentence"`
	BottomContent  string       `db:"bottom_content" json:"bottom_content"`
	Solved         bool         `db:"solved" json:"solved"`
	SolvingContent nulls.String `db:"solving_content" json:"solving_content"`
	SolvingComment nulls.Int    `db:"solving_comment" json:"solving_comment"`
}

type DetailedSentencePost struct {
	ID             int          `db:"id" json:"id"`
	UserID         int          `db:"user_id" json:"user_id"`
	CreatedAt      time.Time    `db:"created_at" json:"created_at"`
	UpdatedAt      time.Time    `db:"updated_at" json:"updated_at"`
	BookID         int          `db:"book_id" json:"book_id"`
	TopContent     string       `db:"top_content" json:"top_content"`
	Sentence       string       `db:"sentence" json:"sentence"`
	BottomContent  string       `db:"bottom_content" json:"bottom_content"`
	Solved         bool         `db:"solved" json:"solved"`
	SolvingContent nulls.String `db:"solving_content" json:"solving_content"`
	SolvingComment nulls.Int    `db:"solving_comment" json:"solving_comment"`
	Rate           nulls.Int    `db:"rate" json:"rate"`
	CommentCount   nulls.Int    `db:"comment_count" json:"comment_count"`
}

func (p *DetailedSentencePost) Split() (Post, SentencePost) {
	return Post{
			ID:        p.ID,
			UserID:    p.UserID,
			CreatedAt: p.CreatedAt,
			UpdatedAt: p.UpdatedAt,
		},
		SentencePost{
			ID:             p.ID,
			BookID:         p.BookID,
			TopContent:     p.TopContent,
			Sentence:       p.Sentence,
			BottomContent:  p.BottomContent,
			Solved:         p.Solved,
			SolvingComment: p.SolvingComment,
		}
}

type DetailedPostComment struct {
	ID        int       `db:"id" json:"id"`
	UserID    int       `db:"user_id" json:"user_id"`
	PostID    int       `db:"post_id" json:"post_id"`
	CreatedAt time.Time `db:"created_at" json:"created_at"`
	UpdatedAt time.Time `db:"updated_at" json:"updated_at"`
	Content   string    `db:"content" json:"content"`
	Rate      nulls.Int `db:"rate" json:"rate"`
}

type FeedPost struct {
	ID        uuid.UUID `db:"id"`
	UserID    int       `db:"user_id"`
	PostID    int       `db:"post_id"`
	CreatedAt time.Time `db:"created_at" json:"created_at"`
	UpdatedAt time.Time `db:"updated_at" json:"updated_at"`
}
